Prototype using the React 16.3 Context API

ghmeier opened the pull request:
[React 16.3](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#official-context-api) introduced a new, official Context API. Here we can use this api to simplify the `BackboneProvider` and the `connectBackboneToReact` function. We no longer need to use soon to be deprecated lifecycle methods and can instead just set up listeners for backbone events and map models to props as previously.

wearhere commented:
> unfortunately, it requires updating to react@^16.3.0 (along with prop-types and react-dom)

It's not possible to check for and conditionally use the new APIs?

I'd be fine with releasing a major version though.

ghmeier commented on lib/backbone-provider.js:
```diff
@@ -1,25 +1,30 @@
+const React = require('react');
+const {Children, Component} = React;
 const PropTypes = require('prop-types');
+const ConnectBackboneToReactContext = require('./context.js'); // eslint-disable-line no-unused-vars
```
Because the file is `.js` so and eslint isn't using the react preset I believe

ghmeier commented:
I've updated this to use the latest `react`, `react-dom`, and `enzyme` versions. As a result, test are fixed. Due to a few changes (apparently) in how these libraries interact, I had to add `wrapper.update()` and re-find the TestComponent in many tests. From what I can tell, this doesn't indicate any functional different. When I checked the actual render functions and the components themselves, they had the correct props/state.

Finally, I removed a failing test that required setState to be called. From what I can tell, we don't actually need setState to be called there since a rerender is triggered already and we don't read from state.

wearhere commented on lib/backbone-provider.js:
```diff
@@ -1,25 +1,35 @@
   render() {
-    return Children.only(this.props.children);
+    return (
+      <ConnectBackboneToReactContext.Provider value={this.state}>
```
Curious would it be equivalent to pass `value={this.props.models}` here (could we do away with `this.state`)?

wearhere commented on lib/connect-backbone-to-react.js:
```diff
@@ -110,33 +99,34 @@ module.exports = function connectBackboneToReact(
-        this.setModels(nextProps, nextContext);
-        this.createNewProps();
+      renderChild({ models } = {}) {
+        const newModels = Object.assign({}, this.models, models, this.props.models);
```
Does this component unregister listeners on any models that are no longer in props? I see it unregister listeners on current models when unmounted, but am not sure it unregisters listeners if the models change while mounted.

I think that the old code did not do this either, but thoughts on fixing this as part of this PR?

yasincanakmehmet replied:
Do we need to remove the listeners? With each renderChild we are creating new models and overriding the previous `this.models`.  I was curious why do we use this.models instead of state. Maybe storing these in the state would be better to trigger re-renders?

wearhere replied:
> Do we need to remove the listeners?

Can we make sure that we're not leaking memory nor having the outdated models trigger unnecessary re-renders (by firing the listeners)?

> I was curious why do we use `this.models` instead of state

I think that the reason that `this.models` is not stored in state is because we instead store [the mapped version of the models](https://github.com/mixmaxhq/connect-backbone-to-react/pull/1/files/3ae64e67ab70a08430ec89c7177eea116ede2dab#diff-ee64377be767f09d5e068f4eb536503dR103) in state although I'm not totally sure about that, they might not be mutually exclusive either. I think some goals there would be a) clarity (of course) and b) that the child component should not see `models` in their props but rather the mapped version that the child has specified.

wearhere commented on lib/context.js:
```diff
@@ -0,0 +1,2 @@
+const React = require('react');
+module.exports = React.createContext('connectBackboneToReact');
```
Perhaps at some point the argument to `createContext` was the name of the context, but [now](https://reactjs.org/docs/context.html#reactcreatecontext) it's a default value of the context. Not sure what that default value should be—`{}` perhaps?

bradvogel commented:
Closing due to age.
